### PR TITLE
[WIP] Ed25519 support for X509 certificates and CSRs

### PR DESCRIPTION
**Pull request by reaperhulk**
depends on #4933

- [ ] _key_identifier_from_public_key
- [ ] docs

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 1]
```diff
@@ -727,11 +727,17 @@ def create_x509_csr(self, builder, private_key, algorithm):
...
+                "MD5 is only (reluctantly) supported for RSA CSRs"
             )
 
         # Resolve the signature algorithm.
-        evp_md = self._evp_md_non_null_from_algorithm(algorithm)
+        if isinstance(private_key, ed25519.Ed25519PrivateKey):
+            if not isinstance(algorithm, hashes.SHA512):
+                raise ValueError("ed25519 requires SHA512 hash algorithm")
+
+            evp_md = self._ffi.NULL
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: This should have a comment explaining that openssl requires null for ed25519 at this time as it always does sha512

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -806,11 +813,17 @@ def create_x509_certificate(self, builder, private_key, algorithm):
...
+                "MD5 is only (reluctantly) supported for RSA certificates"
             )
 
         # Resolve the signature algorithm.
-        evp_md = self._evp_md_non_null_from_algorithm(algorithm)
+        if isinstance(private_key, ed25519.Ed25519PrivateKey):
+            if not isinstance(algorithm, hashes.SHA512):
+                raise ValueError("ed25519 requires SHA512 hash algorithm")
+
+            evp_md = self._ffi.NULL
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Same

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 3]
```diff
@@ -727,11 +727,19 @@ def create_x509_csr(self, builder, private_key, algorithm):
...
             raise ValueError(
-                "MD5 is not a supported hash algorithm for EC/DSA CSRs"
+                "MD5 is only (reluctantly) supported for RSA CSRs"
             )
 
         # Resolve the signature algorithm.
-        evp_md = self._evp_md_non_null_from_algorithm(algorithm)
+        if isinstance(private_key, ed25519.Ed25519PrivateKey):
+            if not isinstance(algorithm, hashes.SHA512):
+                raise ValueError("ed25519 requires SHA512 hash algorithm")
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Add an articl here so this method reads correctly please.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 4]
```diff
@@ -727,11 +727,19 @@ def create_x509_csr(self, builder, private_key, algorithm):
             isinstance(private_key, rsa.RSAPrivateKey)
         ):
             raise ValueError(
-                "MD5 is not a supported hash algorithm for EC/DSA CSRs"
+                "MD5 is only (reluctantly) supported for RSA CSRs"
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: God, can we just kill this?

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 5]
```diff
@@ -806,11 +815,19 @@ def create_x509_certificate(self, builder, private_key, algorithm):
...
-        evp_md = self._evp_md_non_null_from_algorithm(algorithm)
+        if isinstance(private_key, ed25519.Ed25519PrivateKey):
+            if not isinstance(algorithm, hashes.SHA512):
+                raise ValueError("ed25519 requires SHA512 hash algorithm")
+
+            # OpenSSL requires us to pass NULL for EVP_MD at this time, but
+            # will do ed25519ph (SHA512 hashing)
+            evp_md = self._ffi.NULL
+        else:
+            evp_md = self._evp_md_non_null_from_algorithm(algorithm)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Ok, there is now officially too much duplicated code, can we consolidate the common portions of this into a function?

**Pull request closed without merging**